### PR TITLE
Adding a SQL based LeaseManager impl so that we can distribute tasks across a cluster and make sure only 1 node process them at a time.

### PR DESCRIPTION
**Pull request by chirino**
Move the Recordings utility class to it’s own module.

### CODE REVIEW

**Comment by zregvart** [comment]
retest this please

**Comment by zregvart** [comment]
retest this please

**Comment by jimmidyson** [comment]
add to whitelist

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    @PostConstruct
+    public void start() {
+        executor = Executors.newSingleThreadExecutor();
+        scheduler = Executors.newSingleThreadScheduledExecutor();
+        withDB(dbi -> {
+            try {
+                long dbTime = 0;
+
+                // The insert SQL statement used will depend on the DB type that we are using.
```
Comment (KurtStam) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: Two things here:

1. Can we extract the DB specific SQL and put it in the config? This so the code stays DB independent. 
2. I would use the DB as a dumb store and put the logic and time handling in Java. Time is hard.. we have to make sure this works across DST etc. I think there is a hard requirement that time is synced on all nodes on kubernetes.

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    private void doLeaseMaintenance() {
+        withDB(dbi -> {
+            long now = currentTimeMillis();
+
+            // Delete expired leases..
+            dbi.createStatement("DELETE FROM leases where :now > expiration")
+                    .bind("now", now)
+                    .execute();
+
+            // Refresh our leases.
```
Comment (KurtStam) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: Where are the txn boundaries here? I would think there can be conflicts if the same code executes on different nodes with updates?

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            // lets find out which leases we now hold.
+            dbi.createQuery("SELECT lease FROM leases WHERE owner = :owner")
+                    .bind("owner", ownerId)
+                    .map(StringColumnMapper.INSTANCE).forEach(leaseId -> {
+                SqlLease lease = leases.get(leaseId);
+                if (lease != null) {
+                    aquiredLeases.remove(leaseId);
+                    lease.fireAquired();
+                } else {
+                    System.out.println("Weird we have a lease that we did not request");
```
Comment (KurtStam) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: logger?

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                // is too different from the DB time.
+                if (dbTime != 0) {
+                    long now = currentTimeMillis();
+                    long grace = leaseDuration / 4;
+                    if (!(now - grace < dbTime && dbTime < now + (leaseDuration / 4))) {
+                        throw new IllegalStateException("This machine's time is not synchronized with the database.");
+                    }
+                }
+
+            } catch (SQLException e) {
```
Comment (KurtStam) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: log error

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+                            "INSERT into leases (lease, owner, expiration) " +
+                                    "values (:id, :owner, :expiration) " +
+                                    "ON CONFLICT(lease) DO NOTHING";
+
+                    try {
+                        dbTime = dbi.createQuery("select extract(epoch from now()) * 1000")
+                                .map(LongColumnMapper.PRIMITIVE)
+                                .first().longValue();
+                    } catch (Exception e) {
+                        // perhaps we are running agasint cockroachdb..
```
Comment (KurtStam) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: Not sure about this, but hopefully this goes away altogether if we use the db as a simple store.

File: test-support/src/main/java/com/redhat/ipaas/test/Recordings.java [hunk 6]
```diff
@@ -13,18 +13,19 @@
...
 import java.util.List;
 import java.util.concurrent.CountDownLatch;
 
-import javax.annotation.Nonnull;
-
-import org.springframework.cglib.proxy.Enhancer;
-
 /**
  * This class allows you to proxy other objects and record all
  * method invocation done on those objects so that your test cases
```
Comment (KurtStam) [inline test-support/src/main/java/com/redhat/ipaas/test/Recordings.java]: neat :)

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java (hunk 1 above)
Comment (chirino) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: Putting DB specific config is painful for users as we are now pushing this hard task on to them.

The lease manager does generally use the DB as a dumb store, but it depends on all nodes int the cluster to have similar time which is why I added that bit of logic on startup to check to see if the node's  time is in sync with the DB node. We could remove that but then, you could run into cases where thing don't work right if a nodes time is way off.

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java (hunk 5 above)
Comment (chirino) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: which part are you not sure about?

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java (hunk 2 above)
Comment (chirino) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: I think each one of those SQL statements can be executed concurrently and multiple times without issue.  No transactions are needed. BONUS: less locking. 

Can you think of scenario where thats not the case?

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java (hunk 3 above)
Comment (chirino) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: will do.

File: leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java (hunk 4 above)
Comment (chirino) [inline leases/src/main/java/com/redhat/ipaas/leases/impl/SqlLeaseManager.java]: will fix. thx!

**Comment by rhuss** [comment]
Not yet diving in the code, but will it work with JPA ?

**Comment by chirino** [comment]
I'm not sure JPA can do some of the SQL queries being used in this class. Why do you ask?

**Comment by rhuss** [comment]
As said, I have not really yet a clear idea what this PR is supposed to do (not the PRs fault btw ;-), but want to know when using different, JPA based provider for the domain models (so not accessing the DB directly), that then this still works.

**Comment by zregvart** [comment]
I wonder if PostgreSQL advisory locks would be good here (https://www.postgresql.org/docs/9.4/static/explicit-locking.html)

**Comment by chirino** [comment]
@rhuss this does not depend on any of the domain models, so I think we are ok on that front.

**Comment by chirino** [comment]
@zregvart I tried avoiding using features that can only be found in postgresql just in case we ever want to switch off it.

**Comment by zregvart** [comment]
@chirino yeah, I've done that many times, it complicates the implementation and rarely does do the job well and in the end the need to switch to another backend never arises. Anyway just a thought since I started getting these on account of `retest this please` comment :)

**Comment by iocanel** [comment]
retest this

**Comment by KurtStam** [comment]
I'm still concerned about 

1. not having a PENDING state anymore so I don't really understand how the workflows (https://github.com/redhat-ipaas/ipaas-rest/tree/master/docs/workflows) work now and 
2. it still uses the time setting in the db rather then doing it in java code.

**Comment by iocanel** [comment]
retest this please

**Comment by iocanel** [comment]
retest this please

**Comment by KurtStam** [comment]
@chirino @iocanel what is the status of this?

**Comment by jimmidyson** [comment]
Is this still needed or can we close it now? If it's needed, then please rebase, fix merge conflicts and we can re-review? Thanks!

**Comment by chirino** [comment]
closing.. I need to rebase and refresh this.

**Pull request closed without merging**